File: GestionCelulares.Application/Reportes/ReporteService.cs
using GestionCelulares.Application.Common.Interfaces;
using GestionCelulares.Domain.Entities;
using Microsoft.EntityFrameworkCore;

namespace GestionCelulares.Application.Reportes;

public interface IReporteService
{
    Task<ReporteVentasDto> VentasAsync(DateTime desde, DateTime hasta, int? sucursalId);
    Task<ReporteInventarioDto> InventarioAsync(int? sucursalId);
    Task<ReporteMorosidadDto> MorosidadAsync();
    Task<ReporteCajaDto> CajaAsync(DateTime desde, DateTime hasta, int? sucursalId);
    Task<ReporteTallerDto> TallerAsync(DateTime desde, DateTime hasta, int? sucursalId);
    Task<ReporteCobrosDto> CobrosAsync(DateTime desde, DateTime hasta);
}

public class ReporteService : IReporteService
{
    private readonly IApplicationDbContext _db;

    public ReporteService(IApplicationDbContext db) => _db = db;

    public async Task<ReporteVentasDto> VentasAsync(DateTime desde, DateTime hasta, int? sucursalId)
    {
        var fin = hasta.Date.AddDays(1);
        var q = _db.Ventas.AsNoTracking()
            .Where(v => v.Estado == "Completada" && v.Fecha >= desde.Date && v.Fecha < fin);
        if (sucursalId.HasValue)
            q = q.Where(v => v.SucursalId == sucursalId.Value);

        var totales = await q
            .GroupBy(_ => 1)
            .Select(g => new
            {
                Cantidad = g.Count(),
                Subtotal = g.Sum(v => v.Subtotal),
                Descuento = g.Sum(v => v.Descuento),
                Impuesto = g.Sum(v => v.Impuesto),
                Total = g.Sum(v => v.Total)
            })
            .FirstOrDefaultAsync();

        var porDia = await q
            .SelectMany(v => v.Detalles, (v, d) => new
            {
                Fecha = v.Fecha.Date,
                d.Total,
                Ganancia = d.PrecioUnitario * d.Cantidad - d.Descuento
                    - (d.ImeiId != null ? d.Imei!.PrecioCosto : d.Variante.PrecioCosto * d.Cantidad),
                v.VentaId
            })
            .GroupBy(x => x.Fecha)
            .Select(g => new VentaDiaDto
            {
                Fecha = g.Key,
                Cantidad = g.Select(x => x.VentaId).Distinct().Count(),
                Total = g.Sum(x => x.Total),
                Ganancia = g.Sum(x => x.Ganancia)
            })
            .OrderBy(d => d.Fecha)
            .ToListAsync();

        return new ReporteVentasDto
        {
            Desde = desde.Date,
            Hasta = hasta.Date,
            Cantidad = totales?.Cantidad ?? 0,
            Subtotal = totales?.Subtotal ?? 0,
            Descuento = totales?.Descuento ?? 0,
            Impuesto = totales?.Impuesto ?? 0,
            Total = totales?.Total ?? 0,
            Ganancia = porDia.Sum(d => d.Ganancia),
            PorDia = porDia
        };
    }

    public async Task<ReporteInventarioDto> InventarioAsync(int? sucursalId)
    {
        // Costo real desde los IMEIs disponibles
        var costoQ = _db.InventarioImeis.AsNoTracking().Where(i => i.Estado == "Disponible");
        if (sucursalId.HasValue)
            costoQ = costoQ.Where(i => i.SucursalId == sucursalId.Value);

        var costoPorVariante = await costoQ
            .GroupBy(i => new { i.VarianteId, i.SucursalId })
            .Select(g => new { g.Key.VarianteId, g.Key.SucursalId, Costo = g.Sum(i => i.PrecioCosto) })
            .ToListAsync();

        var vistaQ = _db.InventarioDisponible.AsNoTracking();
        if (sucursalId.HasValue)
            vistaQ = vistaQ.Where(x => x.SucursalId == sucursalId.Value);
        var vista = await vistaQ.OrderBy(x => x.Producto).ToListAsync();

        var lineas = vista.Select(x => new InventarioLineaDto
        {
            Producto = x.Producto,
            Marca = x.Marca,
            Color = x.Color,
            Almacenamiento = x.Almacenamiento,
            SucursalId = x.SucursalId,
            Disponibles = x.Disponibles,
            PrecioVenta = x.PrecioVenta,
            ValorCosto = costoPorVariante
                .Where(c => c.VarianteId == x.VarianteId && c.SucursalId == x.SucursalId)
                .Sum(c => c.Costo)
        }).ToList();

        return new ReporteInventarioDto
        {
            Equipos = lineas.Sum(l => l.Disponibles),
            ValorCosto = lineas.Sum(l => l.ValorCosto),
            ValorVenta = lineas.Sum(l => l.PrecioVenta * l.Disponibles),
            Lineas = lineas
        };
    }

    public async Task<ReporteMorosidadDto> MorosidadAsync()
    {
        var clientes = await _db.CuotasVencidas.AsNoTracking()
            .GroupBy(c => new { c.ClienteId, c.Cliente, c.Telefono })
            .Select(g => new MorosidadClienteDto
            {
                ClienteId = g.Key.ClienteId,
                Cliente = g.Key.Cliente,
                Telefono = g.Key.Telefono,
                CuotasVencidas = g.Count(),
                MontoVencido = g.Sum(c => c.Saldo),
                DiasMaxVencido = g.Max(c => c.DiasVencido)
            })
            .OrderByDescending(c => c.MontoVencido)
            .ToListAsync();

        return new ReporteMorosidadDto
        {
            ClientesEnMora = clientes.Count,
            MontoVencidoTotal = clientes.Sum(c => c.MontoVencido),
            Clientes = clientes
        };
    }

    public async Task<ReporteCajaDto> CajaAsync(DateTime desde, DateTime hasta, int? sucursalId)
    {
        var fin = hasta.Date.AddDays(1);
        var q = _db.SesionesCaja.AsNoTracking()
            .Where(s => s.FechaApertura >= desde.Date && s.FechaApertura < fin);
        if (sucursalId.HasValue)
            q = q.Where(s => s.SucursalId == sucursalId.Value);

        var detalle = await q.OrderBy(s => s.FechaApertura)
            .Select(s => new CajaSesionDto
            {
                SesionCajaId = s.SesionCajaId,
                SucursalId = s.SucursalId,
                FechaApertura = s.FechaApertura,
                FechaCierre = s.FechaCierre,
                MontoApertura = s.MontoApertura,
                MontoCierre = s.MontoCierre,
                Diferencia = s.Diferencia,
                Estado = s.Estado
            })
            .ToListAsync();

        return new ReporteCajaDto
        {
            Sesiones = detalle.Count,
            TotalDiferencias = detalle.Sum(s => s.Diferencia ?? 0),
            Detalle = detalle
        };
    }

    public async Task<ReporteTallerDto> TallerAsync(DateTime desde, DateTime hasta, int? sucursalId)
    {
        var fin = hasta.Date.AddDays(1);
        var q = _db.OrdenesTaller.AsNoTracking()
            .Where(o => (o.Estado == "Entregado" && o.FechaEntrega >= desde.Date && o.FechaEntrega < fin)
                     || (o.Estado == "Cancelado" && o.FechaRecepcion >= desde.Date && o.FechaRecepcion < fin));
        if (sucursalId.HasValue)
            q = q.Where(o => o.SucursalId == sucursalId.Value);

        var entregadas = await q.Where(o => o.Estado == "Entregado")
            .Select(o => new
            {
                o.TecnicoId,
                Tecnico = o.Tecnico == null ? "(sin técnico)" : o.Tecnico.NombreCompleto,
                Costo = o.CostoFinal ?? 0,
                o.ComisionTecnico,
                Repuestos = o.Repuestos.Sum(r => (decimal?)(r.Costo * r.Cantidad)) ?? 0
            })
            .ToListAsync();

        var canceladas = await q.CountAsync(o => o.Estado == "Cancelado");

        return new ReporteTallerDto
        {
            Entregadas = entregadas.Count,
            Canceladas = canceladas,
            IngresosPorReparacion = entregadas.Sum(e => e.Costo),
            CostoRepuestos = entregadas.Sum(e => e.Repuestos),
            ComisionesTecnicos = entregadas.Sum(e => e.ComisionTecnico),
            PorTecnico = entregadas
                .GroupBy(e => new { e.TecnicoId, e.Tecnico })
                .Select(g => new TallerTecnicoDto
                {
                    TecnicoId = g.Key.TecnicoId ?? 0,
                    Tecnico = g.Key.Tecnico,
                    OrdenesEntregadas = g.Count(),
                    Ingresos = g.Sum(e => e.Costo),
                    Comision = g.Sum(e => e.ComisionTecnico)
                })
                .OrderByDescending(t => t.Ingresos)
                .ToList()
        };
    }

    public async Task<ReporteCobrosDto> CobrosAsync(DateTime desde, DateTime hasta)
    {
        var fin = hasta.Date.AddDays(1);

        var porDia = await _db.PagosCredito.AsNoTracking()
            .Where(p => p.Fecha >= desde.Date && p.Fecha < fin)
            .GroupBy(p => p.Fecha.Date)
            .Select(g => new CobroDiaDto
            {
                Fecha = g.Key,
                Pagos = g.Count(),
                Total = g.Sum(p => p.Monto)
            })
            .OrderBy(d => d.Fecha)
            .ToListAsync();

        return new ReporteCobrosDto
        {
            Pagos = porDia.Sum(d => d.Pagos),
            TotalCobrado = porDia.Sum(d => d.Total),
            PorDia = porDia
        };
    }
}

File: gestioncelulares-web/src/app/core/reporte.service.ts
import { HttpClient, HttpParams } from '@angular/common/http';
import { Injectable, inject } from '@angular/core';
import { environment } from '../../environments/environment';

export interface ReporteVentas {
  desde: string; hasta: string; cantidad: number;
  subtotal: number; descuento: number; impuesto: number; total: number; ganancia: number;
  porDia: { fecha: string; cantidad: number; total: number; ganancia: number }[];
}

export interface ReporteInventario {
  equipos: number; valorCosto: number; valorVenta: number;
  lineas: { producto: string; marca: string | null; color: string | null; almacenamiento: string | null; sucursalId: number; disponibles: number; precioVenta: number; valorCosto: number }[];
}

export interface ReporteMorosidad {
  clientesEnMora: number; montoVencidoTotal: number;
  clientes: { clienteId: number; cliente: string; telefono: string | null; cuotasVencidas: number; montoVencido: number; diasMaxVencido: number }[];
}

export interface ReporteCaja {
  sesiones: number; totalDiferencias: number;
  detalle: { sesionCajaId: number; sucursalId: number; fechaApertura: string; fechaCierre: string | null; montoApertura: number; montoCierre: number | null; diferencia: number | null; estado: string }[];
}

export interface ReporteTaller {
  entregadas: number; canceladas: number; ingresosPorReparacion: number; costoRepuestos: number; comisionesTecnicos: number;
  porTecnico: { tecnicoId: number; tecnico: string; ordenesEntregadas: number; ingresos: number; comision: number }[];
}

export interface ReporteCobros {
  pagos: number; totalCobrado: number;
  porDia: { fecha: string; pagos: number; total: number }[];
}

export interface Reporte607 {
  periodo: string; rnc: string; cantidad: number;
  totalMontoFacturado: number; totalItbis: number;
  nombreArchivo: string; contenidoTxt: string; sinComprobante: number;
}

export interface Reporte606 {
  periodo: string; rnc: string; cantidad: number;
  totalMontoFacturado: number; totalItbis: number;
  nombreArchivo: string; contenidoTxt: string; sinRnc: number;
}

@Injectable({ providedIn: 'root' })
export class ReporteService {
  private http = inject(HttpClient);
  private base = `${environment.apiUrl}/reportes`;

  private rango(desde: string, hasta: string, sucursalId?: number) {
    let p = new HttpParams().set('desde', desde).set('hasta', hasta);
    if (sucursalId) p = p.set('sucursalId', sucursalId);
    return p;
  }

  ventas(desde: string, hasta: string, sucursalId?: number) {
    return this.http.get<ReporteVentas>(`${this.base}/ventas`, { params: this.rango(desde, hasta, sucursalId) });
  }
  inventario(sucursalId?: number) {
    let p = new HttpParams();
    if (sucursalId) p = p.set('sucursalId', sucursalId);
    return this.http.get<ReporteInventario>(`${this.base}/inventario`, { params: p });
  }
  morosidad() {
    return this.http.get<ReporteMorosidad>(`${this.base}/morosidad`);
  }
  caja(desde: string, hasta: string, sucursalId?: number) {
    return this.http.get<ReporteCaja>(`${this.base}/caja`, { params: this.rango(desde, hasta, sucursalId) });
  }
  taller(desde: string, hasta: string, sucursalId?: number) {
    return this.http.get<ReporteTaller>(`${this.base}/taller`, { params: this.rango(desde, hasta, sucursalId) });
  }
  cobros(desde: string, hasta: string) {
    return this.http.get<ReporteCobros>(`${this.base}/cobros`, { params: this.rango(desde, hasta) });
  }
  reporte607(anio: number, mes: number) {
    const p = new HttpParams().set('anio', anio).set('mes', mes);
    return this.http.get<Reporte607>(`${this.base}/607`, { params: p });
  }
  reporte606(anio: number, mes: number) {
    const p = new HttpParams().set('anio', anio).set('mes', mes);
    return this.http.get<Reporte606>(`${this.base}/606`, { params: p });
  }
}

/** Descarga un archivo de texto plano (ej. el TXT del 607 para la DGII). */
export function descargarTxt(nombre: string, contenido: string): void {
  const blob = new Blob([contenido], { type: 'text/plain;charset=utf-8;' });
  const url = URL.createObjectURL(blob);
  const a = document.createElement('a');
  a.href = url;
  a.download = nombre;
  a.click();
  URL.revokeObjectURL(url);
}

/** Exporta filas a CSV y dispara la descarga. */
export function exportarCsv(nombre: string, columnas: string[], filas: (string | number)[][]): void {
  const escapar = (v: string | number) => {
    const s = String(v ?? '');
    return /[",\n]/.test(s) ? `"${s.replace(/"/g, '""')}"` : s;
  };
  const contenido = [columnas, ...filas].map(f => f.map(escapar).join(',')).join('\n');
  const blob = new Blob(['﻿' + contenido], { type: 'text/csv;charset=utf-8;' });
  const url = URL.createObjectURL(blob);
  const a = document.createElement('a');
  a.href = url;
  a.download = `${nombre}.csv`;
  a.click();
  URL.revokeObjectURL(url);
}
